packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -951,7 +945,24 @@ function use<T>(usable: Usable<T>): T {
       if (thenableState === null) {
         thenableState = createThenableState();
       }
-      return trackUsedThenable(thenableState, thenable, index);
+      const result = trackUsedThenable(thenableState, thenable, index);
+      if (
+        currentlyRenderingFiber.alternate === null &&
+        (workInProgressHook === null
+          ? currentlyRenderingFiber.memoizedState === null
+          : workInProgressHook.next === null)
+      ) {
+        // Initial render, and either this is the first time the component is
+        // called, or there were no Hooks called after this use() the previous
+        // time (perhaps because it threw). Subsequent Hook calls should use the
+        // mount dispatcher.
+        if (__DEV__) {
+          ReactCurrentDispatcher.current = HooksDispatcherOnMountInDEV;
```

Comment:
That dispatcher is related to the hooks mismatch warning, not useDebugValue. But I think I recall the last time I looked at that code, it was due to a weird legacy mode only quirk related to React.lazy. We should try to clean it up (separately). Still, probably the worst thing that happens is we sometimes don't fire a warning where we should.

The other weird case is the InvalidNestedHooksDispatcher(s). If we want to match production exactly then we have to account for those too. That's where you call a hook inside of e.g. useMemo. But that's considered undefined behavior so also not super critical.

So I think we can merge as-is. Neither of those are as important as fixing the bug.

Reply:
Yeah sorry – HooksDispatcherOnMountWithHookTypesInDEV is because non-stateful hooks (like useDebugValue) can trigger the mismatch warning. Some of the tests in ReactHooks-test.internal using useContext or useDebugValue fail without it. But I couldn't get it to happen here with the test I added.

Can you elaborate on what needs to be done with the InvalidNestedHooks ones? From a quick glance I would think we're already OK there and it will warn as expected.

Reply:
The `finally` block when you exit a `useMemo` call sets the dispatcher back to the previous one.

https://github.com/facebook/react/blob/96cdeaf89bfde7551b4ffe4d685da169b780f2f7/packages/react-reconciler/src/ReactFiberHooks.js#L2972-L2976

So for example, if you were to call `use` inside `useMemo` (which triggers a warning in dev), I think what would happen is that the dispatcher would get overridden twice.

Although now that I type that out, maybe it just works because the `useMemo` computation would have been reused from last time. So the invalid hook call wouldn't happen.

Reply:
Meh. I agree it's undefined behavior and you get the first warning so I don't really care.

Reply:
Hmm although now is the dev and prod behavior different…